components/async-container/async-container-mixin.js

```diff
@@ -0,0 +1,90 @@
+
+export const asyncStates = {
+	initial: 'initial',
+	pending: 'pending',
+	complete: 'complete'
+};
+
+export const AsyncContainerMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			asyncPendingDelay: { type: Number, attribute: 'async-pending-delay' },
+			asyncState: { type: String }
+		};
+	}
+
+	constructor() {
+		super();
+		this._initializeAsyncState();
+		this.asyncPendingDelay = 0;
+		if (!this.asyncContainer) {
+			this.addEventListener('pending-state', this._handleAsyncItemState.bind(this));
+		}
+	}
+
+	firstUpdated() {
+		if (!this.asyncContainer) return;
+		const container = this.asyncContainer();
+		if (container) {
+			container.addEventListener('pending-state', this._handleAsyncItemState.bind(this));
+		}
+	}
+
+	resetAsyncState() {
+		this._initializeAsyncState();
+	}
+
+	async _handleAsyncItemState(e) {
+		const promise = e.detail.promise;
+
+		if (!promise) return;
+		if (this.asyncState === asyncStates.complete) return;
+
+		this._asyncPromises.push(promise);
+		this._asyncCounts.pending++;
+
+		if (this.asyncState === asyncStates.initial) {
+			if (this.asyncPendingDelay > 0) {
+				if (this._asyncTimeoutId === null) {
```

Comment:
The gist of this is:

* only go from `initial` to `pending`
* only schedule a delayed state change to pending once

Reply:
This looks correct to me.

Reply:
One simplification might be to still do a timeout even if the value is `0`, so you could remove the `if(this.asyncPendingDelay > 0)` and the `else` branch. Always doing a timeout might always avoid weird race conditions.

Reply:
Yeah sounds good.  I'll switch that.